### ISSUE TITLE
NLog assembly from an embedded resource is still not working

### ISSUE THREAD

**Issue by pimpace** [NLog/NLog#1500]
**Type** (choose one):
- Bug

**NLog version**: (e.g. 4.2.3)
4.4.0-beta11

**Platform**: 
.Net 4.5 

According to the latest change for this part as you tried to check the directory exists, seems not working. :( Im sorry, I cannot provide any exception text, as I just quickly wanted to inform you that the latest release still has problem for this matter. 

On the other hand, I'll try to catch this bug, and will provide a nice exception text...

Keep up the good work!

**Comment by 304NotModified** [comment]
OK please post the exception when you've got it :)

**Comment by pimpace** [comment]
Hey

Sorry for late. I have a great news, seems it can work with embedded state. However...
I use a config file to setup NLog.
Check this out!

```
<?xml version="1.0" encoding="utf-8" ?>
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:schemaLocation="http://www.nlog-project.org/schemas/NLog.xsd NLog.xsd"
      internalLogFile="NLog.log" internalLogLevel="Trace" autoReload="true">

  <targets>
    <target xsi:type="File" name="logfile" fileName="logfile.log"
            layout="${date:format=yyyy-MM-dd HH\:mm\:ss.fff}|${level:uppercase=true}> ${message}"
            deleteOldFileOnStartup="true" />
    <target xsi:type="ColoredConsole" name="console" layout="${date:format=yyyy-MM-dd HH\:mm\:ss.fff}|${level:uppercase=true}> ${message}" >
      <highlight-row condition="level == LogLevel.Info" foregroundColor="DarkGray" />
      <highlight-word regex="(?&lt;=^.{14}).*"  foregroundColor="White" />
    </target>
    <target name="richTextBox_wrapper" xsi:type="AsyncWrapper" queueLimit="5000" overflowAction="Discard" >
      <target xsi:type="RichTextBox" controlName="richTextBoxMain" formName="MainForm"
              autoScroll="true" maxLines="250" allowAccessoryFormCreation="false"  messageRetention="OnlyMissed"
              layout="${date:format=HH\:mm\:ss}|${level:uppercase=true}> ${message}" useDefaultRowColoringRules="false" >

        <!--<row-coloring backgroundColor="Empty" condition="level==LogLevel.Trace" fontColor="White" style="Regular" />-->
        <row-coloring backgroundColor="Empty" condition="level==LogLevel.Debug" fontColor="DarkGray" style="Regular" />
        <row-coloring backgroundColor="Empty" condition="level==LogLevel.Info" fontColor="SteelBlue" style="Regular" />
        <row-coloring backgroundColor="Empty" condition="level==LogLevel.Warn" fontColor="Yellow" style="Regular" />
        <row-coloring backgroundColor="Empty" condition="level==LogLevel.Error" fontColor="Red" style="Bold" />
        <row-coloring backgroundColor="Empty" condition="level==LogLevel.Fatal" fontColor="Magenta" style="Bold" />
      </target>
    </target>
  </targets>
  
  <rules>
    <logger name="Logic.*" minlevel="Trace" writeTo="logfile" />
    <logger name="Program.*" minlevel="Info" writeTo="richTextBox_wrapper" />
  </rules>
</nlog>
```

And now here's the catch. When NLog is embedded, it can't parse the config file which is in its default location, where my executable app is. The parsing method can't reach the config files's end and can't get the rules. Only until <targets> got parsed. I think it cannot able to continue the parsing because of this exception: 
```
2016-06-27 23:17:20.0307 Debug Setting 'AsyncTargetWrapper.queueLimit' to '5000'
2016-06-27 23:17:20.0307 Debug Setting 'AsyncTargetWrapper.overflowAction' to 'Discard'
2016-06-27 23:17:20.0463 Error Error in Parsing Configuration File. Exception: NLog.NLogConfigurationException: Exception occurred when loading configuration from ..\bin\Debug\NLog.config ---> System.ArgumentException: Target cannot be found: 'RichTextBox'
   at NLog.Config.Factory`2.CreateInstance(String name)
   at NLog.Config.XmlLoggingConfiguration.ParseTargetElement(Target target, NLogXmlElement targetElement)
   at NLog.Config.XmlLoggingConfiguration.ParseTargetsElement(NLogXmlElement targetsElement)
   at NLog.Config.XmlLoggingConfiguration.ParseNLogElement(NLogXmlElement nlogElement, String filePath, Boolean autoReloadDefault)
   at NLog.Config.XmlLoggingConfiguration.ParseTopLevel(NLogXmlElement content, String filePath, Boolean autoReloadDefault)
   at NLog.Config.XmlLoggingConfiguration.Initialize(XmlReader reader, String fileName, Boolean ignoreErrors)
   --- End of inner exception stack trace ---
2016-06-27 23:17:20.0619 Error Error has been raised. Exception: System.ArgumentException: Target cannot be found: 'RichTextBox'
   at NLog.Config.Factory`2.CreateInstance(String name)
   at NLog.Config.XmlLoggingConfiguration.ParseTargetElement(Target target, NLogXmlElement targetElement)
   at NLog.Config.XmlLoggingConfiguration.ParseTargetsElement(NLogXmlElement targetsElement)
   at NLog.Config.XmlLoggingConfiguration.ParseNLogElement(NLogXmlElement nlogElement, String filePath, Boolean autoReloadDefault)
   at NLog.Config.XmlLoggingConfiguration.ParseTopLevel(NLogXmlElement content, String filePath, Boolean autoReloadDefault)
   at NLog.Config.XmlLoggingConfiguration.Initialize(XmlReader reader, String fileName, Boolean ignoreErrors)
2016-06-27 23:17:20.0619 Debug --- NLog configuration dump ---
2016-06-27 23:17:20.0619 Debug Targets:
2016-06-27 23:17:20.0619 Debug File Target[logfile]
2016-06-27 23:17:20.0619 Debug ColoredConsole Target[console]
2016-06-27 23:17:20.0619 Debug Rules:
2016-06-27 23:17:20.0619 Debug --- End of NLog configuration dump ---


```
As you can see here, parser couldn't get the _row-coloring_ lines.
"Target cannot be found: 'RichTextBox'"

I don't get it why. At _ParseTargetElement(Target target, NLogXmlElement targetElement)_ function seems crashed. Due to this: _xsi:type="RichTextBox"_

But what could be the difference when it runs from existed file?
Is it related to _NLog.Windows.Forms_?

I hope these info could help you! Tell me if some other tests, info, anything need!

**Comment by 304NotModified** [comment]
hey!

I think the automatic registration of the extension (NLog.Windows.Forms) won't work with a embedded config file. That's the reason you get "Target cannot be found: 'RichTextBox'".

You need the ` <extensions> ` syntax, see https://github.com/NLog/NLog/wiki/Configuration-file#extensions

The last error should not interrupt your program, but if it does, then please update to the latest version (4.3.5)

**Comment by 304NotModified** [comment]
NB: probable the 4.3.x version is working better for the full .NET platform. 4.4 is still beta

**Comment by pimpace** [comment]
Yo!

I got it Mate! 

Sorry, maybe I wasn't clear enough, you've misunderstood me. I'm embedding all my _.dll_ I use into my _.exe_ I don't embed the _NLog.config_ file. 

But I got the problem:
As I wanna use this great logger for a RichTextBox control, I use _NLog.Windows.Forms_ extension to get RichTextBox target property access. And I found this:

```
            var extensionDlls = Directory.GetFiles(assemblyLocation, "NLog*.dll")
                .Select(Path.GetFileName)
                .Where(x => !x.Equals("NLog.dll", StringComparison.OrdinalIgnoreCase))
                .Where(x => !x.Equals("NLog.UnitTests.dll", StringComparison.OrdinalIgnoreCase))
                .Where(x => !x.Equals("NLog.Extended.dll", StringComparison.OrdinalIgnoreCase))
                .Select(x => Path.Combine(assemblyLocation, x));

```
This is in BuildDefaultFactory() (NLog.Config.ConfigurationItemFactory)

As _NLog.Windows.Forms.dll_ is also an embedded resource you cannot able to get it with a file access. (Directory.GetFiles) :)

**Comment by 304NotModified** [comment]
Indeed. But you can work around it right?

**Comment by pimpace** [comment]
Yep, trying to find a solution when _ConfigurationItemFactory_ get all the resources (*.dll's) then load them for all the config properties accesses.

**Comment by pimpace** [comment]
LOL

It was simplest as we can get! :)

As you pointed out with <extensions> syntax, my problem was that _NLog.Windows.Forms.dll_ hasn't loaded so we got an exception as we don't have access the RichTextBox target. I use Costura.Fody to embed all my references. So only things I needed to do is put there lines into NLog.config:
```
    <extensions>
      <add assembly="NLog.Windows.Forms"/>
    </extensions>
```
NLog checks extensions and access it, so Costura unzips and loads them. That's it. Thanks all your help Mate! I use the latest beta of NLog, so I don't know if it'd work with latest stable. Will try and get back here to tell.
